### ISSUE TITLE
Plated einsum() interface is awkward

### ISSUE THREAD

**Issue by fritzo** [pyro-ppl/pyro#1508]
Pyro's generalized [einsum()](http://docs.pyro.ai/en/dev/ops.html#pyro.ops.contract.einsum) interface is awkward in a few ways. We should fix these before moving upstream to [opt_einsum](https://github.com/dgasmith/opt_einsum) ~~and before Pyro's upcoming 0.3 release~~.

## Issue 1: `einsum()` doesn't handle all inputs

Plated einsum problems are either polynomial or exponential in the sizes of input tensors, and the boundary is subtle and not obvious from looking at an equation string (requires search for certain subgraphs). Currently plated `einsum()` implements only the polynomial solution, and raises `NotImplementedError` in case of exponential complexity. Is this ok? Should we instead warn and attempt a solution?

## Issue 2: plated `einsum()` does too much work

Naive plated `einsum()` does more work than is required in applications of tensor message passing. As of #1507 we instead implement `einsum(..., modulo_total=True)` which avoids passing messages amongst unrelated batches; the default `modulo_total=False` raises a `NotImplementedError`, so users must opt-in to this behavior. Note that it is subtle to determine whether batches are related, and is not obvious from looking at an equation string (requires connected component analysis). Is there a cleaner way to define the operation we want to perform?

## Issue 3: inferring batch context

Currently plated `einsum()` infers the batch context of each sum dimension by examining the minimum context of each dim. This is problematic because `einsum()` is not decomposable, e.g. we can't split `u("ai,ai,a->",x,y,z,plates="i")` into `u("a,a->",u("ai,ai->a",x,y,plates="i"),z)`. Two possible resolutions are 
1. Replace the `plates` kwarg with a dict mapping sum-dims to their set of batch dims.
    **Pros:** unambiguous
    **Cons:** verbose
2. Place a restriction on dim ordering such that each sum dim is batched by a batch dim iff it is to the right of the batch dim, e.g. the above example would work, but `u("ia,ia,a->",...)` would error.
    **Pros:** unambiguous, succinct
    **Cons:** more restrictive, requires users to `.permute()` inputs

**Comment by fritzo** [comment]
@srush we welcome design suggestions for the `ubersum()` interface.

**Comment by martinjankowiak** [comment]
regarding issue 1, it seems to me that we don't want to make exponential time stuff always fail, since the computation might conceivably be exponential in a smallish number. perhaps by default fail but have some sort of `allow_exponential_computation` flag or the like?

regarding issue 3, my initial reaction is to be on board with option 2, although i still need to think about this some more.

**Comment by eb8680** [comment]
I agree with Martin re: issue 1, and I think the `modulo_total` flag is a reasonable solution to 2.

Re: 3, what do you think about using the type of `batch_dim` to dispatch between options 1 and 2, calling 1 when it's a dict, and constructing a dict and implementing option 2 as a call to option 1 when it's a string?

**Comment by fritzo** [comment]
@srush The only reference we've published is the [ubersum docstring](http://docs.pyro.ai/en/dev/ops.html#pyro.ops.contract.ubersum). Whereas `einsum()` can be used to implement variable elimination on a discrete graphical model, `ubserum()` is the natural generalization to graphical models with plates, i.e. batches of independent random variables. If you end up using `ubersum()`, we'd be happy to hear about any issues you have.

**Comment by srush** [comment]
(This maybe belongs in a different thread)

There is a common case in NLP where the model is almost a UGM but you need a bit more flexibility to implement it efficiently. In particular for models like HSMM or parsing, I want to be able to sum over vectors of the same size. For instance "d,e,f,da,eb,fc" and then I want to sum g = a+b+c. I guess you could do something like "d,e,f,da,eb,fc,abch->h", for some constant "abch", but I think that is inefficient? Is there an easy way to do this that I am missing?

**Comment by eb8680** [comment]
@srush `ubersum` allows multiple outputs, so you should be able to use `"d,e,f,da,eb,fc->a,b,c"` and sum the results.  Is that what you're looking for?

**Comment by srush** [comment]
No I mean. I want the sum within the expression so I can chain it forward. 
ubersum("d,e,f,dh,eh,fh,hi,.....->p", sum_dim="h")

Maybe I'm missing the point, but what seems neat to me is that ubersum is a domain specific language for specifying dynamic programs over a semiring. However handling "branches" i.e. sums requires leaving the language.

**Comment by fritzo** [comment]
Hi @srush and @justinchiu,

@eb8680 and I have been working on a new Pyro backend for dynamic programming and nonstandard interpretation. We'd love to hear your feedback on the [design doc](https://docs.google.com/document/d/1NVlfQnNQ0Aebg8vfIGcJKsnSqAhB4bbClQrb5dwm2OM/edit?usp=sharing), both because you've expressed interest in more complex dynamic programs, and because we've copied some of your design choices from [namedtensor](https://github.com/harvardnlp/namedtensor).

Using our [prototype interface](https://github.com/pyro-ppl/funsor), I believe your above example could be written as something like:
```py
import funsor, torch
from funsor import Tensor, bint
u = Tensor(torch.tensor(...), OrderedDict([('d', bint(10)]))
v = Tensor(torch.tensor(...), OrderedDict([('e', bint(11)]))
w = Tensor(torch.tensor(...), OrderedDict([('f', bint(12)]))
x = Tensor(torch.tensor(...), OrderedDict([('d', bint(10)), ('a', bint(20))]))
y = Tensor(torch.tensor(...), OrderedDict([('e', bint(11)), ('a', bint(20))]))
z = Tensor(torch.tensor(...), OrderedDict([('f', bint(12)), ('a', bint(20))]))
with funsor.interpretation(my_interpretation):
    result = (u * x).sum('d') + (v * y).sum('e') + (w * z).sum('f')
    # ...do other stuff with +, *, .sum(), .prod(), ...
```
(sorry if I've again butchered your example; the language is quite general and I'm confident we could express what you were thinking). We can then use nonstandard interpretation to do things like: optimize the expression tree before evaluating, or replace `.sum()` reduction over real free variables by Monte Carlo sampling.

Note that our aim is not a "higher-level PyTorch" but rather a "lower-level Pyro" :smile:

**Comment by fritzo** [comment]
Closing since [funsor](https://github.com/pyro-ppl/funsor) seems to be a better solution for general variable elimination.
